### PR TITLE
Tax clean-up and refactoring (SHOOP-975)

### PR DESCRIPTION
**Pull request by suutari-ai**
Main part of all this goes to refactoring everything to use
Shop.create_price.  That should prevent taxful/taxless price mismatches
in a Shop.

Prices and other money amounts are now passed as Money and Price objects
in the code, which ensures some more type safety. Yeppee!

TaxModule API was rethought through and got some changes.

OrderSource, BaseBasket and StoredBasket now have the price
unit (=currency+taxfullness) stored in them to avoid mixing units after
a basket load.

Calculation of taxes is made more explicit in OrderSource API.

Made OrderLine, SourceLine and PriceInfo to use unified API for price
fields (named Priceful).  This should help when implementing template
helpers for rendering prices taxful/taxless depending on
user's/merchant's preference (Shop's taxfullness could differ from
rendering taxfullnes as long as TaxModule may calculate taxes
efficiently, which of course is not always true though.)

There is some small fixes and refactorings included here and there for
some things I found out while implementing this.  Hope you don't mind.

TODO:
 * Assigning customers to correct tax groups on creation (retail
   customers and company customers)
 * Implementing value-added taxes correctly (i.e. not just selecting the
   first tax, but actually adding them together)
 * The taxful/taxless price rendering stuff as said above.
 * Cleaning up the TODOs
 * Documentation of the whole system (for developers and for merchants)

Good luck reviewing! Have fun!

### CODE REVIEW

File: shoop/admin/currencybound.py [hunk 1]
```diff
@@ -0,0 +1,22 @@
...
+# Copyright (c) 2012-2015, Shoop Ltd. All rights reserved.
+#
+# This source code is licensed under the AGPLv3 license found in the
+# LICENSE file in the root directory of this source tree.
+
+from shoop.core.models import Shop
+
+
+class CurrencyBound(object):
+    def __init__(self, currency=None, *args, **kwargs):
```
Comment (akx) [inline shoop/admin/currencybound.py]: Docstring? What's the use for this class?

File: shoop/admin/modules/orders/dashboard.py [hunk 2]
```diff
@@ -45,25 +58,31 @@ def get_subtitle(count):
     return _("Based on %d orders") % count
 
 
-def get_sales_of_the_day_block(request):
+def get_sales_of_the_day_block(currency, request):
```
Comment (akx) [inline shoop/admin/modules/orders/dashboard.py]: `request` is usually the first argument, I'd appreciate it if it was the same here.

File: shoop/core/fields/__init__.py [hunk 3]
```diff
@@ -52,12 +53,17 @@ def deconstruct(self):
         return (name, path, args, kwargs)
 
 
-class MoneyField(models.DecimalField):
+class CurrencyField(models.CharField):
```
Comment (akx) [inline shoop/core/fields/__init__.py]: Even if the field itself might not have `choices`, it should generate a select-type `formfield` by default. That can be added later though.

File: shoop/core/models/contacts.py [hunk 4]
```diff
@@ -85,7 +85,10 @@ class Meta:
 
 class CompanyContact(Contact):
     members = models.ManyToManyField("Contact", related_name="company_memberships", blank=True)
-    vat_code = models.CharField(max_length=32, blank=True, verbose_name=_('VAT code'))
+    tax_number = models.CharField(
+        max_length=32, blank=True,
```
Comment (akx) [inline shoop/core/models/contacts.py]: Should this length be 64 too, as it is for `Address`es?

File: shoop/core/models/orders.py [hunk 5]
```diff
@@ -175,7 +180,7 @@ class Order(models.Model):
         "Address", related_name='shipping_orders',  blank=True, null=True,
         on_delete=models.PROTECT,
         verbose_name=_('shipping address'))
-    vat_code = models.CharField(max_length=20, blank=True, verbose_name=_('VAT code'))
+    tax_number = models.CharField(max_length=20, blank=True, verbose_name=_('Tax number'))
```
Comment (akx) [inline shoop/core/models/orders.py]: The length should be 64 here too, I suppose.

File: shoop/core/models/orders.py [hunk 6]
```diff
@@ -214,9 +219,15 @@ class Order(models.Model):
...
-    taxful_total_price = MoneyField(editable=False, verbose_name=_('grand total'), default=0)
-    taxless_total_price = MoneyField(editable=False, verbose_name=_('taxless total'), default=0)
-    display_currency = models.CharField(max_length=4, blank=True)
+    taxful_total_price = TaxfulPriceProperty('taxful_total_price_value', 'currency')
+    taxless_total_price = TaxlessPriceProperty('taxless_total_price_value', 'currency')
+
+    taxful_total_price_value = MoneyValueField(editable=False, verbose_name=_('grand total'), default=0)
+    taxless_total_price_value = MoneyValueField(editable=False, verbose_name=_('taxless total'), default=0)
+    currency = CurrencyField()
+    prices_include_tax = models.BooleanField()  # TODO: (TAX) Document Order.prices_include_tax
```
Comment (akx) [inline shoop/core/models/orders.py]: But we already have fields for both taxful and taxless prices here.. Why is this required?

File: shoop/core/models/products.py [hunk 7]
```diff
@@ -243,7 +242,7 @@ def get_shop_instance(self, shop):
         if cached:
             return cached
 
-        shop_inst = self.shop_products.filter(shop=shop).first()
+        shop_inst = self.shop_products.get(shop=shop)
         if shop_inst:
```
Comment (akx) [inline shoop/core/models/products.py]: As the line above is `.get()`, this `if` becomes unnecessary.

File: shoop/admin/currencybound.py (hunk 1 above)
Comment (suutari-ai) [inline shoop/admin/currencybound.py]: I'll add..

File: shoop/core/pricing/_priceful_properties.py [hunk 8]
```diff
@@ -0,0 +1,68 @@
...
+        :rtype: TaxlessPrice
+        """
+        if price.includes_tax:
+            inv_tax_ratio = (taxless.value / taxful.value if taxful else 1)
+            return TaxlessPrice(price.amount * inv_tax_ratio)
+        else:
+            return price
+
+
+class Alias(object):
```
Comment (akx) [inline shoop/core/pricing/_priceful_properties.py]: Wouldn't this belong in `shoop.utils.properties`?

File: shoop/core/pricing/priceful.py [hunk 9]
```diff
@@ -0,0 +1,207 @@
...
+
+    taxful_discounted_unit_price = TaxfulFrom('discounted_unit_price')
+    taxless_discounted_unit_price = TaxlessFrom('discounted_unit_price')
+
+    taxful_unit_discount_amount = TaxfulFrom('unit_discount_amount')
+    taxless_unit_discount_amount = TaxlessFrom('unit_discount_amount')
+
+    # TODO: Replace occurences of these aliases with their source fields
+
+    total_price = Alias('price')
```
Comment (akx) [inline shoop/core/pricing/priceful.py]: Maybe this sort of "TODO"ed alias should emit a Warning when `__get__`ed?

File: shoop/admin/modules/orders/dashboard.py (hunk 2 above)
Comment (suutari-ai) [inline shoop/admin/modules/orders/dashboard.py]: OK, I'll change..

File: shoop/core/fields/__init__.py (hunk 3 above)
Comment (suutari-ai) [inline shoop/core/fields/__init__.py]: I may need some help with that. For now I just wanted the CurrencyFields to be separated from general purpose CharFields.

File: shoop/core/models/contacts.py (hunk 4 above)
Comment (suutari-ai) [inline shoop/core/models/contacts.py]: Yes, I had a todo that "make all the tax number fields equal length". Didn't do that yet, as you can see..

File: shoop/front/checkout/single_page.py [hunk 10]
```diff
@@ -30,9 +33,9 @@ def _to_choices(objects):
     return [(x.id, x) for x in objects]
 
 
-class OrderForm(forms.Form):
+class OrderForm(forms.Form, TaxNumberCleanMixin):
```
Comment (akx) [inline shoop/front/checkout/single_page.py]: Don't mixins usually go on the left?

File: shoop/core/fields/__init__.py (hunk 3 above)
Comment (akx) [inline shoop/core/fields/__init__.py]: Sure, no worries. As mentioned, it can be done later, and adding a `formfield` method to a class won't cause new migrations.

File: shoop/core/models/orders.py (hunk 6 above)
Comment (suutari-ai) [inline shoop/core/models/orders.py]: Because order.taxful_total_price is a TaxfulPrice object, but order.taxful_total_price_value is a Decimal. Operating with Prices makes it more type safe and allows smarter rendering of this stuff in the templates (since Price has the taxfulness info and the currency).

Or what did you mean?

File: shoop/core/pricing/_priceful_properties.py (hunk 8 above)
Comment (suutari-ai) [inline shoop/core/pricing/_priceful_properties.py]: I want to get rid of it very soon, so I'd rather see it deleted than moved..

File: shoop/utils/i18n.py [hunk 11]
```diff
@@ -43,9 +40,19 @@ def get_current_babel_locale():
     return get_babel_locale(locale_string=translation.get_language())
 
 
-def format_home_currency(value, locale=None):
-    value = parse_decimal_string(value)
-    return format_currency(value, currency=settings.SHOOP_HOME_CURRENCY, locale=locale or get_current_babel_locale())
+def format_money(amount, digits=None, widen=0, locale=None):
+    loc = babel.Locale.parse(locale or get_current_babel_locale())
```
Comment (akx) [inline shoop/utils/i18n.py]: Use `get_babel_locale()` (above)?

File: shoop/core/pricing/priceful.py (hunk 9 above)
Comment (suutari-ai) [inline shoop/core/pricing/priceful.py]: Or we could just search-replace all the `(.*line).total_price` occurences and replace them with `\1.price`?

File: shoop/utils/i18n.py [hunk 12]
```diff
@@ -43,9 +40,19 @@ def get_current_babel_locale():
...
+
+    pattern = loc.currency_formats.get(None).pattern
+
+    # pattern is a formatting string.  Couple examples:
+    # '¤#,##0.00', '#,##0.00\xa0¤', '\u200e¤#,##0.00', and '¤#0.00'
+
+    if digits is not None:
+        pattern = pattern.replace(".00", "." + (digits * "0"))
+    if widen:
```
Comment (akx) [inline shoop/utils/i18n.py]: Won't `digits` and `widen` used together have a somewhat strange effect?

File: shoop/utils/properties.py [hunk 13]
```diff
@@ -0,0 +1,178 @@
...
+    tranform passed kwargs to the fields specified in the
+    `MoneyProperty`.
+    """
+    def __init__(self, *args, **kwargs):
+        transformed = _tranform_init_kwargs(type(self), kwargs)
+        super(MoneyPropped, self).__init__(*args, **kwargs)
+        _check_transformed_types(self, transformed)
+
+
+def _tranform_init_kwargs(cls, kwargs):
```
Comment (akx) [inline shoop/utils/properties.py]: Typo: tranform?

File: shoop/utils/properties.py [hunk 14]
```diff
@@ -0,0 +1,178 @@
...
+    value_class = TaxlessPrice
+
+
+class MoneyPropped(object):
+    """
+    Mixin for transforming MoneyProperty init parameters.
+
+    Add this mixin as (first) base for the class that has
+    `MoneyProperty` properties and this will make its `__init__`
+    tranform passed kwargs to the fields specified in the
```
Comment (akx) [inline shoop/utils/properties.py]: Typo: tranform?

File: shoop/utils/properties.py [hunk 15]
```diff
@@ -0,0 +1,178 @@
...
+    for field in list(kwargs.keys()):
+        prop = getattr(cls, field, None)
+        if isinstance(prop, MoneyProperty):
+            value = kwargs.pop(field)
+            _tranform_single_init_kwarg(prop, field, value, kwargs)
+            transformed.append((field, value))
+    return transformed
+
+
+def _tranform_single_init_kwarg(prop, field, value, kwargs):
```
Comment (akx) [inline shoop/utils/properties.py]: Typo: tranform?

**Comment by akx** [comment]
Aside from the few nits I found to pick, this looks like good stuff.

I know you're diligent enough to have tested the logic thoroughly, so I didn't start to build a "mental map" of it right now -- especially with a diff this huge.

File: shoop/core/models/orders.py (hunk 6 above)
Comment (akx) [inline shoop/core/models/orders.py]: I meant -- why is the concrete field of taxes being included needed on the model, as we have fields for both taxful and taxless prices anyway?

I mean, if we have taxful_total_price = 1.1, taxless_total_price = 1.0... what would prices_include_tax be? :D

File: shoop/core/pricing/priceful.py (hunk 9 above)
Comment (akx) [inline shoop/core/pricing/priceful.py]: Sure. Just figured that'd be more automatic in a sense. But I guess we can add the warning feature during that cleanup pass.

**Comment by Pikkupomo** [comment]
Issue with TaxRule
```
Rule 1:
Tax = 25% tax
customer_tax_group = "normal customer"
country_code_pattern = FI
postal_code_pattern=20300-20400

Rule 2:
Tax = 50% tax
country_code_pattern = FI
postal_code_pattern=20300-20400

Customer has Tax group = "normal customer"
```
In checkout confirm, you see 50% taxation in use. this surely is a bug:
image

**Comment by Pikkupomo** [comment]
Patterns are utterly broken in many ways and missing help texts. Not sure if this is in scope of this PR but wanted to mention

**Comment by Pikkupomo** [comment]
If I change a tax class for a product that someone has in basket, the frontend of the shop will be in limbo and error `Conflicting product and line tax classes: <TaxClass:2-> vs <TaxClass:1-default>` is shown.

**Comment by Pikkupomo** [comment]
Saving Tax object for the second time will cause problem:
* ImmutabilityError at /sa/taxes/tax/1/ Tax objects are immutable

Understandable we want to limit modifying taxes after adding, but only if it hasn't been used or something. (i made a typo, now i cannot fix a typo)

**Comment by Pikkupomo** [comment]
Breadcrumb `(Home Taxes Finnish normal taxation)` the `taxes` points to "tax classes"

File: shoop/core/models/products.py (hunk 7 above)
Comment (Pikkupomo) [inline shoop/core/models/products.py]: also wont this fail miserably if there are multiple shops and products for them?

File: shoop/front/checkout/single_page.py (hunk 10 above)
Comment (suutari-ai) [inline shoop/front/checkout/single_page.py]: You're right. Good catch. I'll fix..

File: shoop/utils/i18n.py (hunk 11 above)
Comment (suutari-ai) [inline shoop/utils/i18n.py]: Gotta see if that works.. Can't remember any more why it was done this way..

File: shoop/utils/i18n.py (hunk 12 above)
Comment (suutari-ai) [inline shoop/utils/i18n.py]: No, it would just add them together, i.e. digits=3, widen=4 -> 7 digits. IMHO there's no need to restrict that even though I wouldn't expect anyone to combine them.

File: shoop/utils/properties.py (hunk 13 above)
Comment (suutari-ai) [inline shoop/utils/properties.py]: Oops, I'll fix.

File: shoop/utils/properties.py (hunk 14 above)
Comment (suutari-ai) [inline shoop/utils/properties.py]: Seemz that I have a problem with that letter, maybe it'z broken on my keyboard..

**Comment by akx** [comment]
Changes LGTM, @suutari-ai :)

File: shoop/core/models/orders.py (hunk 6 above)
Comment (suutari-ai) [inline shoop/core/models/orders.py]: I thought that would be misleading.. that's the reason for the todo note. ;-)

Order.prices_include_tax is for the OrderLines, since their values have to be stored pre-tax-calculation, we need the boolean so we know how to calculate the taxes for the lines (upwards or downwards from the values stored in the lines). Does that make any sense?

**Comment by suutari-ai** [comment]
retest this please

**Comment by akx** [comment]
Well, then.
